### DISCUSSION TITLE
How to propagate ansi codes in tables through a pipe?

### DISCUSSION THREAD

**Discussion by alecandido** [nushell/nushell#12780] [Q&A] [answered]
It seems that every time a command is targeting standard output, ansi codes are removed.

https://github.com/nushell/nushell/blob/460a1c8f874b503cb4140619838be483600f00ce/crates/nu-command/src/viewers/table.rs#L965

This is usually a good default, but sometimes you might want to preserve them...

```nu
config env --default | nu-highlight | lines | less -R
```

**Comment by fdncred** [comment]
I don't have issues when doing it this way `config env --default | nu-highlight | less`. I also have an env var set like this `$env.LESS = "-FRXS"`

**Reply by alecandido** [reply]
Ok, you're definitely right, so the question is wrong. It is not always stripping codes for every output.

Still, the example with `lines` is colorless. And that is consistent with the cited line, because I didn't notice that the code was scoped in the `table.rs` module (I found it searching, I haven't paid enough attention to the context :facepalm:).

**Reply by ghost** [reply, marked as answer]
i found a workaround: `| to text | less`

it somehow determines if the it ever reaches a external command and then strips the ansi (checkable with `| lines | into string | save -r foo.txt; cat foo.txt`, etc).  
it seems to forget that its the same data when using `| to text`. i **assume** it stores some sort of "this is ansi" inside of the string and to text clears the metadata.

**Reply by alecandido** [reply]
@Jan9103 you're right that some formatting is preserved (and in particular the ansi codes related to the coloring introduced by `nu-highlight`), but in the original example it just reverts the `lines` command, without which there is no issue at all, as @fdncred observed.

Instead, if you try with multi-columns tables (e.g. `ls | to text | less`), you'll find out that the result is definitely far from the desired ones (not only the ansi codes are lost, but it is just formatted as the textual representation of the individual records, joint with `\n`).

**Reply by ghost** [reply]
it still works, you just have to add table if you want the expanded form: `config env --default | nu-highlight | lines | table | to text | less`.
at least for me it works and puts highlighted text in a line-enumerated format into less (altough id just use `| bat` usually)

**Reply by alecandido** [reply]
Ok, you're right: `ls | table | to text | less` is exactly working as expected as well.

So, `| table | to text | less` is the full workaround.

> at least for me it works and puts highlighted text in a line-enumerated format into less (altough id just use `| bat` usually)

I also use `| bat`, and even more often `| nvim -`. But I found that suggestion while looking for something else, and it was frustrating not to be able to propagate the ansi codes...

Thanks for the answer!
